reflect/src/main/java/dagger/reflect/Binding.java

```diff
@@ -115,33 +109,31 @@
       Annotation[] methodAnnotations = method.getDeclaredAnnotations();
       Annotation qualifier = findQualifier(methodAnnotations);
       Key dependency = Key.of(qualifier, method.getReturnType());
-      return new Key[] { dependency };
+      return new LinkRequest(new Key[] { dependency }, new boolean[] { true });
     }
 
     @Override
-    public Binding<Optional<T>> link(Binding<?>[] dependencies) {
-      return new LinkedOptionalBinding<>(dependencies);
+    public LinkedBinding<?> link(LinkedBinding<?>[] dependencies) {
+      return new LinkedOptionalBinding<>(dependencies[0]);
     }
   }
 
   final class LinkedOptionalBinding<T> extends LinkedBinding<Optional<T>> {
-    private final Binding<?>[] dependencies;
+    private @Nullable LinkedBinding<?> dependency;
 
-    public LinkedOptionalBinding(Binding<?>[] dependencies) {
-      this.dependencies = dependencies;
+    LinkedOptionalBinding(@Nullable LinkedBinding<?> dependency) {
+      this.dependency = dependency;
     }
 
     @Override
     public Optional<T> get() {
-      Binding<?> dependency = dependencies[0];
-      if (dependency == null) {
-        return Optional.empty();
-      }
-      return Optional.of((T) dependency.get());
+      return dependency == null
+          ? Optional.empty()
+          : Optional.of((T) dependency.get());
```

Comment:
`dependency.get()` can be `null` (you were wondering here: https://github.com/JakeWharton/dagger-reflect/pull/21/commits/e3d9abd255edfc639110e483113d747775cfce34#r253727983):
```
java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:203)
	at java.util.Optional.<init>(Optional.java:96)
	at java.util.Optional.of(Optional.java:108)
	at dagger.reflect.Binding$LinkedOptionalBinding.get(Binding.java:132)
```
codegen throws: `java.lang.NullPointerException: Cannot return null from a non-@Nullable @Provides method`
to repro change `OptionalBinding` test to:
```java
@Provides static String foo() { return sneaky(); }
static String sneaky() { return null; }
```

Reply:
Yes but it's existing behavior. Should be fixed separately.